lib/ansible/cli/__init__.py

```diff
@@ -253,7 +254,9 @@ def ask_passwords():
         except EOFError:
             pass
 
-        return (sshpass, becomepass)
+        # we 'wrap' the passwords to prevent templating as
+        # they can contain special chars and trigger it incorrectly
+        return (wrap_var(sshpass), wrap_var(becomepass))
```

Comment:
These passwords are byte strings, and `UnsafeProxy` and by extension `wrap_var` doesn't support byte strings.  Your fix would maybe work on py2, but wouldn't work on py3.  Currently on py2 it would be converting bytes to text, which is not what we want.

I've wanted to add bytes functionality to `UnsafeProxy` before, but ended up not doing so.  I think we'll need something like the following added:

```diff
diff --git a/lib/ansible/utils/unsafe_proxy.py b/lib/ansible/utils/unsafe_proxy.py
index 2a180e6d4a..50ff4096af 100644
--- a/lib/ansible/utils/unsafe_proxy.py
+++ b/lib/ansible/utils/unsafe_proxy.py
@@ -53,8 +53,8 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
-from ansible.module_utils.six import string_types, text_type
-from ansible.module_utils._text import to_text
+from ansible.module_utils.six import binary_type, string_types, text_type
 from ansible.module_utils.common._collections_compat import Mapping, MutableSequence, Set
 
 
@@ -69,14 +69,15 @@ class AnsibleUnsafeText(text_type, AnsibleUnsafe):
     pass
 
 
+class AnsibleUnsafeBytes(binary_type, AnsibleUnsafe):
+    pass
+
+
 class UnsafeProxy(object):
     def __new__(cls, obj, *args, **kwargs):
-        # In our usage we should only receive unicode strings.
-        # This conditional and conversion exists to sanity check the values
-        # we're given but we may want to take it out for testing and sanitize
-        # our input instead.
-        if isinstance(obj, string_types):
-            obj = to_text(obj, errors='surrogate_or_strict')
+        if isinstance(obj, binary_type):
+            return AnsibleUnsafeBytes(obj)
+        elif isinstance(obj, text_type):
             return AnsibleUnsafeText(obj)
         return obj
 ```

It should be noted however, that this will change behavior on py2 elsewhere in the code, so we may need to audit what may break due to this change.

Reply:
a less intrusive change (and use class directly and not wrap_var in this PR), specially for backporting, then we can look at migrating to the more general and 'correct' version above 

```diff --git a/lib/ansible/utils/unsafe_proxy.py b/lib/ansible/utils/unsafe_proxy.py
index 2a180e6d4a..d1d2296fee 100644
--- a/lib/ansible/utils/unsafe_proxy.py
+++ b/lib/ansible/utils/unsafe_proxy.py
@@ -53,7 +53,7 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type

-from ansible.module_utils.six import string_types, text_type
+from ansible.module_utils.six import string_types, text_type, binary_type
 from ansible.module_utils._text import to_text
 from ansible.module_utils.common._collections_compat import Mapping, MutableSequence, Set

@@ -68,6 +68,9 @@ class AnsibleUnsafe(object):
 class AnsibleUnsafeText(text_type, AnsibleUnsafe):
     pass

+class AnsibleUnsafeBytes(binary_type, AnsibleUnsafe):
+    pass
+

 class UnsafeProxy(object):
     def __new__(cls, obj, *args, **kwargs):
@@ -75,9 +78,8 @@ class UnsafeProxy(object):
         # This conditional and conversion exists to sanity check the values
         # we're given but we may want to take it out for testing and sanitize
         # our input instead.
-        if isinstance(obj, string_types):
-            obj = to_text(obj, errors='surrogate_or_strict')
-            return AnsibleUnsafeText(obj)
+        if isinstance(obj, string_types) and not isinstance(AnsibleUnsafeBytes):
+            obj = AnsibleUnsafeText(to_text(obj, errors='surrogate_or_strict'))
         return obj```

Reply:
You still need to explicitly do something with `AnsibleUnsafeBytes`.

Reply:
other than use on the passwords?

Reply:
Sorry, I see that in your fix, but it was not included in your above diff.

Reply:
no, i stated that in the comment, but didnt put it in code (had not edited file yet)